[fullscreen] Use a single queue for event dispatching

nt1m opened the pull request:
#### 0b476fcbb873dbc513bdc6ec486c0fc69097b1f4
<pre>
[fullscreen] Use a single queue for event dispatching
<a href="https://bugs.webkit.org/show_bug.cgi?id=288292">https://bugs.webkit.org/show_bug.cgi?id=288292</a>
<a href="https://rdar.apple.com/145372389">rdar://145372389</a>

Reviewed by Alex Christensen.

<a href="https://fullscreen.spec.whatwg.org/#run-the-fullscreen-steps">https://fullscreen.spec.whatwg.org/#run-the-fullscreen-steps</a> specifies a single queue should be used.

This slightly changes the order of change events relative to the error events.

Other stylistic changes:
- Merge some helper methods back into `dispatchPendingEvents`, since they&apos;re only used once after 289984@main.
- Rename variables to match spec terminology
- Simply check the event type instead of having a separate `shouldNotifyMediaElement` bool argument

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-ready-check-not-allowed-cross-origin.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-cross-origin.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/remove-last-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::dispatchPendingEvents):
(WebCore::FullscreenManager::queueFullscreenChangeEventForDocument):
(WebCore::FullscreenManager::dispatchEventForNode): Deleted.
(WebCore::FullscreenManager::dispatchFullscreenChangeOrErrorEvent): Deleted.
(WebCore::FullscreenManager::emptyEventQueue): Deleted.
* Source/WebCore/dom/FullscreenManager.h:

Canonical link: <a href="https://commits.webkit.org/290898@main">https://commits.webkit.org/290898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4cf5da3f301a875487bed8f67b6718d01b9fc5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91422 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10954 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/463 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96390 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42111 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93472 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19296 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/27716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94423 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/82809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/50523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41280 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98393 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18584 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18839 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/78647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/78422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/22942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11716 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14462 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18582 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23858 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112755 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18294 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/112755 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21752 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20059 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -111,7 +111,7 @@ void FullscreenManager::requestFullscreenForElement(Ref<Element>&& element, Full
         ERROR_LOG(identifier, message);
         if (emitErrorEvent == EmitErrorEvent::Yes) {
-            m_fullscreenErrorEventTargetQueue.append(WTFMove(element));
+            m_pendingEvents.append(std::pair { EventType::Error, WTFMove(element) });
```
I had to specify `std::pair`, otherwise it wouldn't let me `WTFMove` a `Ref` into `GCReachableRef`. Happy to take any suggestions for improvements if anyone has any.

achristensen07 commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -619,55 +619,46 @@ void FullscreenManager::dispatchPendingEvents()
-}
+    // Steps 1-2:
+    PendingEventsList pendingEvents;
+    m_pendingEvents.swap(pendingEvents);
```
I think this would be better like this:
auto pendingEvents = std::exchange(m_pendingEvents, { });

nt1m committed:
[fullscreen] Use a single queue for event dispatching
https://bugs.webkit.org/show_bug.cgi?id=288292
rdar://145372389

Reviewed by Alex Christensen.

https://fullscreen.spec.whatwg.org/#run-the-fullscreen-steps specifies a single queue should be used.

This slightly changes the order of change events relative to the error events.

Other stylistic changes:
- Merge some helper methods back into `dispatchPendingEvents`, since they're only used once after 289984@main.
- Rename variables to match spec terminology
- Simply check the event type instead of having a separate `shouldNotifyMediaElement` bool argument

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-ready-check-not-allowed-cross-origin.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-cross-origin.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/remove-last-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::dispatchPendingEvents):
(WebCore::FullscreenManager::queueFullscreenChangeEventForDocument):
(WebCore::FullscreenManager::dispatchEventForNode): Deleted.
(WebCore::FullscreenManager::dispatchFullscreenChangeOrErrorEvent): Deleted.
(WebCore::FullscreenManager::emptyEventQueue): Deleted.
* Source/WebCore/dom/FullscreenManager.h:

Canonical link: https://commits.webkit.org/290898@main
